bootstore/src/schemes/v0/state_initial_member.rs

```diff
@@ -0,0 +1,411 @@
+                self.resolve_learn_requests(&rack_secret);
+            output.envelopes.extend_from_slice(&envelopes);
+            output.persist = persist;
+        }
+        output
+    }
+
+    // We've just recomputed the rack secret. Now resolve any pending learn
+    // requests. Return whether we need to persist state and any envelopes to
+    // send to peers.
+    fn resolve_learn_requests(
+        &mut self,
+        rack_secret: &RackSecret,
+    ) -> (bool, Vec<Envelope>) {
+        let mut persist = false;
+        let mut envelopes =
+            Vec::with_capacity(self.pending_learn_requests.len());
+        match self.pkg.decrypt_shares(rack_secret) {
+            Ok(shares) => {
+                while let Some((to, metadata)) =
+                    self.pending_learn_requests.pop_first()
+                {
+                    let Output {
+                        persist: persist_once,
+                        envelopes: envelopes_once,
+                        ..
+                    } = send_share_response(
+                        to.clone(),
+                        metadata.request_id,
+                        &self.pkg,
+                        &mut self.distributed_shares,
+                        &shares,
+                    );
+
+                    if persist_once {
+                        persist = true;
+                    }
+                    envelopes.extend(envelopes_once);
+                }
+            }
+            Err(_) => {
+                envelopes = self.resolve_learn_requests_with_error(
+                    Error::FailedToDecryptShares,
+                );
+            }
+        }
+
+        (persist, envelopes)
+    }
+
+    /// An error prevented successful resolution of learn requests
+    pub fn resolve_learn_requests_with_error(
+        &mut self,
+        error: Error,
+    ) -> Vec<Envelope> {
+        let mut envelopes =
+            Vec::with_capacity(self.pending_learn_requests.len());
+        while let Some((to, metadata)) = self.pending_learn_requests.pop_first()
+        {
+            let msg = Response {
+                request_id: metadata.request_id,
+                type_: error.into(),
+            };
+            envelopes.push(Envelope { to, msg: msg.into() })
+        }
+        envelopes
+    }
+}
+
+impl StateHandler for InitialMemberState {
+    fn handle_request(
+        mut self,
+        common: &mut FsmCommonData,
+        from: Baseboard,
+        request_id: Uuid,
+        request: RequestType,
+    ) -> (State, Output) {
+        use RequestType::*;
+        let output = match request {
+            Init(new_pkg) => {
+                if new_pkg == self.pkg {
+                    // Idempotent response given same pkg
+                    Output::respond(from, request_id, ResponseType::InitAck)
+                } else {
+                    let rack_uuid = self.pkg.rack_uuid;
+                    Output::respond(
+                        from,
```

Comment:
Vague API Vibes: the fact that `from` and `request_id` are passed in here, and then used in every call to `Output::respond` makes me wonder whether `Output::response` should return a smaller type, then the caller should staple `from` and `request_id` onto it.

We'd still need to pass `from` in, because it's used by `decrypt_and_send_share_response`, but could avoid a bit of repetition and plumbing.

Reply:
No longer relevant as of [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)